pkg/ledger/ledger.go

```diff
@@ -0,0 +1,182 @@
+	if opIdxItem == nil {
+		opIdxItem, err = createNewProviderIndexEntry(parsedURI, opIdx)
+		if err != nil {
+			return err
+		}
+	}
+
+	// get key ledger index
+	pklIdx, err := getPklIndex(opIdxItem.Path)
+	if err != nil {
+		return err
+	}
+
+	// get active keys to detect key rotation
+	expectedActiveJWKs := make(map[string]PklIndexItem)
+	for id, jwkIdx := range pklIdx.Items {
+		if jwkIdx.Status == "active" {
+			expectedActiveJWKs[id] = jwkIdx
+		}
+	}
+
+	// query jwks_uri and record time
+	jwks, timestamp, err := getJWKS(parsedURI)
+	if err != nil {
+		return err
+	}
+	if cfg.GetString("loglevel") == "debug" {
+		json, err := jsonStructToString(jwks)
+		if err != nil {
+			return err
+		}
+		log.Debugf(json)
+		log.Debugf("timestamp=%d", timestamp)
+	}
+
+	// update JWK ledger files based on JWKS response from OP
+	pklIdxUpdated := false
+	for _, jwk := range jwks.Keys {
+		// check if JWK already exists in ledger
+		if jwkInLedger(jwk, pklIdx) {
+			// reconcile active jwk to detect key rotation (active key not in JWKS response)
+			reconcileExpectedJWK(jwk, expectedActiveJWKs)
+
+			// TODO check configuration parameter for fail-safe updates
+			// then update exp time based on JWKS query timestamp if true
+			continue
+		}
+		// write new jwk ledger file
+		pklID := hashString(jwk.Kid)
+		newPklFile := PklFile{
+			Jwk: jwk.RawJSON,
+			Nbf: &timestamp,
+			Exp: nil,
+		}
+		ledgerFilePath := filepath.Join(LedgerPath, parsedURI.Host, pklID+LedgerFileExt)
+		err = writeJSONFile(ledgerFilePath, newPklFile)
+		if err != nil {
+			return err
+		}
+
+		// add JWK to ledger index
+		newPklIndexItem := PklIndexItem{
+			Status: StatusActive,
+			Path:   ledgerFilePath,
+		}
+		pklIdx.Items[jwk.Kid] = newPklIndexItem
+		pklIdxUpdated = true
+	}
+
+	// detect rotated keys
+	pklIdxUpdated, err = detectRotatedJWK(pklIdx, expectedActiveJWKs, timestamp)
+	if err != nil {
+		return err
+	}
+
+	// write ledger index updates if modified
+	if pklIdxUpdated {
+		err = writeJSONFile(opIdxItem.Path, pklIdx)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func detectRotatedJWK(pklIdx PklIndex, expectedActiveJWKs map[string]PklIndexItem, timestamp int64) (bool, error) {
+	if len(expectedActiveJWKs) > 0 {
+		log.Infof("remaining active JWKs: %d", len(expectedActiveJWKs))
+		// key was rotated set exp and update ledger index
+		for id, rotatedJWK := range expectedActiveJWKs {
+			// set exp for jwk
+			var jwk PklFile
+			err := readJSONFile(rotatedJWK.Path, &jwk)
+			if err != nil {
+				return false, err
+			}
+			jwk.Exp = &timestamp
+			err = writeJSONFile(rotatedJWK.Path, jwk)
+			if err != nil {
+				return false, err
+			}
+
+			// update ledger index status
+			indexItem, ok := pklIdx.Items[id]
+			if ok {
+				indexItem.Status = StatusArchived
+				pklIdx.Items[id] = indexItem
+			}
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func reconcileExpectedJWK(jwk JWK, activeJWKs map[string]PklIndexItem) {
```

Comment:
I would inline this function to make it more obvious that the map is being mutated

Reply:
[ca3ef02](https://github.com/mrjoelkamp/tuf-pk-ledger/pull/5/commits/ca3ef02e83ba631630cbf021f60a324533ed4b33)